src/hierarchical.rs

```diff
@@ -0,0 +1,493 @@
+
+        let a = replace(&mut dendrograms[leaf], None).unwrap();
+        let b = replace(&mut dendrograms[merge], None).unwrap();
+
+        if leaf < merge {
+            dendrograms[merge] = Some(Box::new(Dendrogram::Branch(lambda[leaf], a, b)));
+        } else {
+            dendrograms[leaf] = Some(Box::new(Dendrogram::Branch(lambda[leaf], b, a)));
+        }
+    }
+
+    dendrograms.swap_remove(points.len() - 1).unwrap()
+}
+
+
+#[cfg(test)]
+#[allow(non_snake_case)]
+fn minimal_distance(d: &Fn(usize, usize) -> f64, A: &[usize], B: &[usize]) -> f64 {
+    A.iter().flat_map(|&a|
+        B.iter().map(move |&b| Dist(d(a,b)))
+    ).min().unwrap().0
+}
+
+#[allow(non_snake_case)]
+fn maximal_distance(d: &Fn(usize, usize) -> f64, A: &[usize], B: &[usize]) -> f64 {
+    A.iter().flat_map(|&a|
+        B.iter().map(move |&b| Dist(d(a,b)))
+    ).max().unwrap().0
+}
+
+#[derive(PartialEq, PartialOrd)]
+struct Dist(f64);
+impl Eq for Dist { }
+impl Ord for Dist {
+    fn cmp(&self, other: &Self) -> Ordering {
+        self.partial_cmp(other).unwrap()
+    }
+}
+
+
+/// A hierarchical tree structure describing the merge operations.
+#[derive(Clone, Debug)]
+pub enum Dendrogram<T> {
+    /// The union of two sub-trees (clusters) and their distance.
+    Branch(f64, Box<Dendrogram<T>>, Box<Dendrogram<T>>),
+
+    /// The leaf node contains an element.
+    Leaf(T)
+}
+
+impl<T: PartialEq> PartialEq for Dendrogram<T> {
+    fn eq(&self, other: &Self) -> bool {
+        if let (&Dendrogram::Leaf(ref a), &Dendrogram::Leaf(ref b)) = (self, other) {
+            a == b
+        } else if let (&Dendrogram::Branch(d1, ref a1, ref b1),
+            &Dendrogram::Branch(d2, ref a2, ref b2)) = (self, other)
+        {
+            d1 == d2 && ((a1 == a2 && b1 == b2) || (a1 == b2 && a2 == b1))
+        } else {
+            false
+        }
+    }
+}
+
+
+/// An `Iterator` over all elements contained in an `Dendrogram`.
+#[derive(Clone, Debug)]
+pub struct Elements<'a, T: 'a> {
+    items: Vec<&'a Dendrogram<T>>
+}
+
+impl<'a, T> Iterator for Elements<'a, T> {
+    type Item = &'a T;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        self.items.pop().and_then(|item| {
+            match item {
+                &Dendrogram::Branch(_, ref a, ref b) => {
+                    self.items.push(a);
+                    self.items.push(b);
+
+                    self.next()
+                },
+                &Dendrogram::Leaf(ref p) => Some(p)
+            }
+        })
+    }
+}
+
+impl<'a, T> IntoIterator for &'a Dendrogram<T> {
+    type Item = &'a T;
+    type IntoIter = Elements<'a, T>;
+
+    fn into_iter(self) -> Self::IntoIter {
+        Elements {
+            items: vec![self]
+        }
+    }
+}
+
+
+#[cfg(test)]
+mod tests {
+    use permutohedron::Heap;
+    use quickcheck::{Arbitrary, Gen, TestResult, quickcheck};
+
+    use Euclid;
+    use super::{Agglomerative, LinkageCriterion, minimal_distance};
+
+    #[test]
+    fn single() {
```

Comment:
For sure. Although, it would be good to have one or two tests with fixed input data, to catch, for example, the different implementations having the same bug (e.g. if they both make the same incorrect assumptions about the input format or whatever).